pkg/operator/controllers/checkers/internetchecker/controller.go

```diff
@@ -0,0 +1,119 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	checkercommon "github.com/Azure/ARO-RP/pkg/operator/controllers/checkers/common"
+	"github.com/Azure/ARO-RP/pkg/util/conditions"
+)
+
+// This is the permissions that this controller needs to work.
+// "make generate" will run kubebuilder and cause operator/deploy/staticresources/*/role.yaml to be updated
+// from the annotation below.
+// +kubebuilder:rbac:groups=aro.openshift.io,resources=clusters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=aro.openshift.io,resources=clusters/status,verbs=get;update;patch
+
+const (
+	ControllerName = "InternetChecker"
+)
+
+// Reconciler runs a number of checkers
+type Reconciler struct {
+	log  *logrus.Entry
+	role string
+
+	arocli  aroclient.Interface
+	checker internetChecker
+}
+
+func NewReconciler(log *logrus.Entry, arocli aroclient.Interface, role string) *Reconciler {
+	return &Reconciler{
+		log:  log,
+		role: role,
+
+		arocli:  arocli,
+		checker: newInternetChecker(),
+	}
+}
+
+// Reconcile will keep checking that the cluster can connect to essential services.
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !instance.Spec.OperatorFlags.GetSimpleBoolean(checkercommon.ControllerEnabled) {
+		r.log.Info("controller is disabled")
+		return reconcile.Result{}, nil
+	}
+
+	r.log.Info("running")
+	checkErr := r.checker.Check(instance.Spec.InternetChecker.URLs)
+	var condition *operatorv1.OperatorCondition
+	if checkErr != nil {
+		condition = &operatorv1.OperatorCondition{
+			Type:    r.conditionType(),
+			Status:  operatorv1.ConditionFalse,
+			Message: checkErr.Error(),
+			Reason:  "CheckFailed",
+		}
+	} else {
+		condition = &operatorv1.OperatorCondition{
+			Type:    r.conditionType(),
+			Status:  operatorv1.ConditionTrue,
+			Message: "Outgoing connection successful",
+			Reason:  "CheckDone",
+		}
+	}
+
+	err = conditions.SetCondition(ctx, r.arocli, condition, r.role)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// We always requeue here:
+	// * Either immediately (with rate limiting) based on the error
+	//   when checkErr != nil.
+	// * Or based on RequeueAfter when err == nil.
+	return reconcile.Result{RequeueAfter: time.Hour}, checkErr
+}
+
+func (r *Reconciler) conditionType() string {
+	switch r.role {
+	case "master":
+		return arov1alpha1.InternetReachableFromMaster
+	case "worker":
+		return arov1alpha1.InternetReachableFromWorker
+	default:
+		r.log.Warnf("unknown role %s, assuming worker role", r.role)
+		return arov1alpha1.InternetReachableFromWorker
```

Comment:
do we just want to return `"UnknownRole"` here? I think defaulting to `worker` is fine as this should be a weird corner case to start with.

Reply:
We should not hit this as long as we have this in place:

https://github.com/Azure/ARO-RP/blob/b45b3a1d6d2c21acddb866920dd327dc224fde52/cmd/aro/operator.go#L56-L59

However I'm not a big fan of implicit things and I would love to `NewReconciler` to return an error if internet checker controller is being constructed with a role other than master or worker. For rest of the controllers we will only support master role so their `NewReconciler` should return an error for anything other than master.

I will try to do that once I finish the split (assuming I'll have a bit of time before I need to pick up my next card).

But I agree with your other comment: since we have this behaviour - it should be tested. I updated the test to do that (and thanks to you found an issue with the test).

Also your comment reminded me that we should probably set `Unknown` condition status in case if the controller is disabled.

Reply:
Awesome, I agree with the controller enforcing which roles it deals with too. If you run out of time LMK, I'd be happy to pick it up.